addons/purchase/models/purchase.py

```diff
@@ -1045,7 +1045,8 @@ def write(self, values):
     def _unlink_except_purchase_or_done(self):
         for line in self:
             if line.order_id.state in ['purchase', 'done']:
-                raise UserError(_('Cannot delete a purchase order line which is in state \'%s\'.') % (line.state,))
+                state_description = {state_desc[0]: state_desc[1] for state_desc in self._fields['state']._description_selection(self.env)}
```

Comment:
Create the state dict at the beginning of the method, to avoid to call it several times.